### PR TITLE
Data: Add GenericFileWriterFactory

### PR DESCRIPTION
**Pull request by aokolnychyi**
This PR adds `GenericFileWriterFactory`, similar to `FlinkFileWriterFactory` and `SparkFileWriterFactory`. This is a new API that should be used in favor of methods in `FileAppenderFactory` for creating data, equality and position delete writers. Spark, for instance, has migrated to `FileWriterFactory` a long time ago.

This PR migrates all tests to use this API for writing test files and adds a new test suite.

### CODE REVIEW

File: flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestMetadataTableReadableMetrics.java [hunk 1]
```diff
@@ -219,27 +219,27 @@ public void testPrimitiveColumns() throws Exception {
 
     Row binaryCol =
         Row.of(
-            59L,
```
Comment (aokolnychyi) [inline flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestMetadataTableReadableMetrics.java:222]: The column sizes have changed cause the new writer picks up all table properties, which was not true before. I validated the actual values are correct using `parquet-tools`.

**Comment by aokolnychyi** [comment]
@nastra @Fokko @flyrain @amogh-jahagirdar, could you check this one?

File: data/src/test/java/org/apache/iceberg/io/TestFileWriterFactory.java [hunk 2]
```diff
@@ -76,18 +76,15 @@ public static Object[] parameters() {
...
 
   private StructLike partition = null;
   private OutputFileFactory fileFactory = null;
+  private List<T> dataRows;
 
   public TestFileWriterFactory(FileFormat fileFormat, boolean partitioned) {
     super(TABLE_FORMAT_VERSION);
     this.fileFormat = fileFormat;
     this.partitioned = partitioned;
-    this.dataRows =
```
Comment (aokolnychyi) [inline data/src/test/java/org/apache/iceberg/io/TestFileWriterFactory.java:88]: Had to move the initialization after the table creation.

File: flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestMetadataTableReadableMetrics.java (hunk 1 above)
Comment (szehon-ho) [reply flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestMetadataTableReadableMetrics.java:222]: I probably should have originally picked these up from the file itself, but its a suggestion for another PR

**Review by szehon-ho — APPROVED** [review]
It looks good to me.  Whose the audience of this, is it generics?  Still in early stages of re-ramping up and so may miss some context

**Comment by aokolnychyi** [comment]
@szehon-ho, I faced some test failures in other PRs because of issues in `GenericFileAppenderFactory`. It was not that easy to fix that class so I went ahead and added `GenericFileWriterFactory`.

File: flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestMetadataTableReadableMetrics.java (hunk 1 above)
Comment (aokolnychyi) [reply flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestMetadataTableReadableMetrics.java:222]: Yep, we should do that eventually.

**Comment by aokolnychyi** [comment]
Thanks for reviewing, @szehon-ho!

**Pull request merged**